### ISSUE TITLE
Discuss creation of setup.sh 

### ISSUE THREAD

**Issue by lucamlouzada** [gentzkow/GentzkowLabTemplate#26]
This issue is part of an effort to implement substantive improvements to the lab template, as discussed in https://github.com/gentzkow/GentzkowLabTemplate/issues/16.

I am creating this issue to keep track of our discussion on whether to bring back `setup.sh` as suggested e.g in https://github.com/gentzkow/GentzkowLabTemplate/issues/24#issuecomment-2412592632, https://github.com/gentzkow/GentzkowLabTemplate/issues/20#issuecomment-2384365857, and https://github.com/gentzkow/GentzkowLabTemplate/issues/20#issuecomment-2409942947. 

A possible strategy might be to have a `setup.sh` script in the lib that:
- Creates `local_env.sh` if it does not exist
- Prompts the user to edit `local_env.sh `to fill in the relevant paths
- Creates `/external/ `with the appropriate symlinks if these were specified in `local_env.sh`
- Populates each module with a `make.sh` and a `get_inputs.sh` if these don't exist (this would reduce the need of keeping several copies of these in the template)
- Runs `get_inputs.sh` in each module
- Runs check_setup.sh 
- Maybe: Installs any R / Python / Stata dependencies we want to install in the "simple" repo case (no conda or Renv) (we might not want this, see https://github.com/gentzkow/GentzkowLabTemplate/issues/20#issuecomment-2384509608

The one thing I want to flag is the different behaviors that we want depending on whether this script would be run for the first time in a module or not. For example, we might not want it to run `get_inputs.sh` if it's run by the first time (as there won't be any valid input paths); and we might not want it to re-populate the modules and overwrite changes to `local_env.sh ` or `make.sh` that users have already done. I think this is easily handled by some conditionals that identify whether it's the first time it's being run.

I am assigning myself to this, but would like @gentzkow's inputs as we make this decision. Thanks!

**Comment by gentzkow** [comment]
Thanks @lucamlouzada! This sounds good to me. Quick thoughts.

1. Let's discuss "Prompts the user to edit `local_env.sh` to fill in the relevant paths" before we implement it. I'd like to make sure we're aligned on how to do this. Obviously the main thing is to prompt them to fill in paths to the externals.

2. We should definitely only create `get_inputs.sh`, `local_env.sh`, and `make.sh` if they don't already exist. To be on the safe side, we could prompt the user before doing this -- maybe one prompt for whether to create `local_env.sh` and another for whether to populate the modules with `get_inputs.sh` and `make.sh`.

3. We should be able to make the default `get_inputs.sh` be robust to the case where there aren't yet any inputs to get.

**Comment by lucamlouzada** [comment]
Hi @gentzkow ,

I have started writing a draft of this `setup.sh` script but ran into some issues that made me rethink about what we actually want to achieve with this. The main contribution of this script would be to populate the modules with `make.sh` and `get_inputs.sh`. The other things (`local_env.sh` and any eventual dependency checks) can easily be done by `check_setup.sh`. 

The problem with populating the modules is that it requires a strict definition of what constitutes a module, for the script to identify which folders to populate. Currently, part of the [definition](https://github.com/gentzkow/GentzkowLabTemplate/wiki#modules) of a module is a folder that has a `make.sh` script. We could use something like a folder that has a "source" folder instead. However, I feel like this might limit the flexibility of the template for users or lead to unexpected behaviors (what if they have a subfolder in one of their data directories in `raw` that is called "source"?). I'm not sure there is a lot of gain from not having to copy the example modules which already have a source folder and a `make.sh` script as users do currently. 

Is the idea that users should run `setup.sh` only once right after cloning the template? Or that they can run it for the first time after creating their project modules, and running again whenever they create new modules? 

If they are supposed to run it often, I'm not sure there is a gain vs the current approach (which does not require running code after creating new modules).

If they should run it just once, perhaps it's better to embrace the idea that `setup.sh` is a helper script for setting up the template with all the examples - we could then think about including in it things like already populating the example input folders, checking the examples work, etc. It would be something that helps us (as the developers of this template) not have to make changes to several `make.sh` scripts as often, and helps users run the examples if they want to do so, but it would not be something that people are required to run to use the template. In this case, the example template wouldn't even have to have all the modules, as they could be created by this "setup_examples.sh" script. It could have just two modules for example, "0_raw" and "1_analysis". Then, if people want to see how the examples work, they could run "setup_examples.sh" and it would populate the template as it currently is with all the 4 modules, and could already copy the input files and everything.

**Comment by gentzkow** [comment]
@lucamlouzada All good points.

I agree with you that we were creating confusion by mixing two things:

1. A setup script that a user would run when they first clone the **template**, before building their own repository. This would involve things like populating directories with `make.sh`. 
2. A setup script that a user would run when they first clone an already built **repository**. This would involve things like creating `local-env.sh`. 

I agree that (1) is better handled by a `setup_examples.sh` script. Let's set that aside for now and revisit it in a separate issue.

We can then redefine this issue to focus on (2). Let's imagine a user who clones an existing repository like `selective-exposure` or `SearchMarket` and wants to set it up for the first time. I think that setup should involve the following.
* Create `local_env.sh` and populate it with appropriate paths, executable names, etc.
* Populate `/externals/` at the root of the repo with symlinks
* Populate individual `/input/` directories in the various modules with symlinks

I agree that this could all be handled by `check_setup.sh`. As an aesthetic matter, though, I like the idea of having the script that you run to **set up** the repository be called `setup.sh` and the script that you run to **check whether your setup is correct** be called `check_setup.sh`. The question is then how best to integrate those. One version is that `check_setup.sh` just looks to see whether `local_env.sh` and `/externals/` exist and if not it says whats missing then aborts whatever is running and prompts the user to go run `setup.sh`.

**Comment by lucamlouzada** [comment]
Thanks @gentzkow !
I have added the creation of `setup_examples.sh` to the to do list in #16. I also added a note about `check_setup.sh` which was already in the to do list. I'll create issues for these once we have closed out the ones that are currently open.

In b2f32f8, I pushed a version of [`setup.sh`](https://github.com/gentzkow/GentzkowLabTemplate/blob/b2f32f894a2a7b668f80d1a25ab231fe4e4d1c5e/lib/setup/setup.sh) that does what you suggested. It's not going to run right now because it depends on `get_inputs.sh` and `make_externals.sh` which will be created in the PRs for #20 and #24, but I wanted to push it so you can take a look and see if that's the behavior you are envisioning.

The code starts by checking if `local_env.sh` exists. If not, it creates it from the template and prompts the user:
```diff
if [ ! -f "${REPO_ROOT}/local_env.sh" ]; then 
    echo "The file local_env.sh was not found at the root of the repository."
    echo "I will create it from /lib/setup/local_env_template.sh."
    echo "Please edit this file to reflect your local executables and external file paths."
    cp "${REPO_ROOT}/lib/setup/local_env_template.sh" "${REPO_ROOT}/local_env.sh"
fi
```

Then, it loops through the folders in the repo root to find the modules. Folders that have a `make.sh` file are considered a module. In each module, it runs `get_inputs.sh`. This is robust to a module not having a `get_inputs.sh` file, and in #20 I have made it so that `get_inputs.sh` is robust to file paths being invalid or not having any file paths.

Finally, it runs `make_externals.sh` as an optional feature. Note that if a user clones SearchMarket for the first time, for example, they would have to run `setup.sh` without the external links first. Then, after `local_env.sh` is created, they can edit it and add their external paths, and then they would run `setup.sh` again. 

It would be nice for `setup.sh` to run `check_setup.sh` at the end, but I think we could add this after we have worked on `check_setup.sh` which will be in a different issue.

**Comment by gentzkow** [comment]
@lucamlouzada Thanks! This definitely looks like the right idea.

1. Could we simplify the search for the modules and just run all scripts in the repo called `get_inputs.sh`?

2. I think we should make the script interactive so the user affirms they want to do each step. E.g., after saying `local_env.sh` doesn't exist, we should ask the user if they want to create it from `local_env_template.sh`. If they say no we stop the script and ask them to re-run it once that file exists. 

3. Once we create `local_env.sh` from the template, what about pausing the script and asking the user to go edit the paths. Then they can either press continue or re-run `setup.sh` once they've done so.

4. Can we have the externals step uncommented but just have it not create `/externals/` if there are no externals defined in `local_env.sh`?

**Comment by lucamlouzada** [comment]
Thanks @gentzkow , these are all great points. I have made these changes in 5bc5c5a. The script now prompts users before creating `local_env.sh` (or tells them if it already exists). It also prompts the user once before running all `get_inputs.sh` - let me know if you'd prefer it prompts once for each module. It does not prompt the user before creating the external folder, but if there are no external links specified in `local_env.sh`, it will not create the folder (and will tell the user that no external links were found). I made this last change in 7b32e1a for the issue branch in #20.

**Comment by gentzkow** [comment]
Great! Looking good. 

* Let's put `setup.sh` at the root of the repo
* Let's modify the user prompts so they don't assume the user knows what `local_env.sh`, `get_inputs.sh`, etc. are -- i.e., we can give a bit more explanation
* I'm currently getting some errors:

> Please choose: Would you like to create local_env.sh from /lib/setup/local_env_template.sh? (y/n): 
> y
> local_env.sh has been created from the template.
> Please edit it to reflect your local executables and external file paths.
> Press Enter to continue after editing or re-run setup.sh once you've done so.
> setup.sh:read:21: -p: no coprocess

and

> Please choose: Would you like to run all get_inputs.sh scripts in the repository? (y/n): 
> y
> 
> All get_inputs.sh scripts have been run.
> /Users/gentzkow/Dropbox (Personal)/git/GentzkowLabTemplate/lib/shell/make_externals.sh:source:46: no such file or directory: /Users/gentzkow/Dropbox (Personal)/git/GentzkowLabTemplate/lib/shell/make_externals.sh

**Comment by lucamlouzada** [comment]
Thanks @gentzkow ! I have made these changes.

> Let's put setup.sh at the root of the repo

Done in 695df60 and 3eb57e4.

> Let's modify the user prompts so they don't assume the user knows what local_env.sh, get_inputs.sh, etc. are -- i.e., we can give a bit more explanation

Done in 46f27c6. I added short message before each prompt.

> I'm currently getting some errors:
setup.sh:read:21: -p: no coprocess

This error was due to zsh and bash inconsistencies. I pushed a change in a3340fb and it should now be working.

> I'm currently getting some errors:
/Users/gentzkow/Dropbox (Personal)/git/GentzkowLabTemplate/lib/shell/make_externals.sh:source:46: n

This is expected as the file `make_externals.sh` does not exist, it will be created in the PR for #20. However you raise a good point that the `setup.sh` script should be robust to errors of this kind. I added error handling in 2bea120 in the spirit of #19. 

As soon as we have implemented the changes in the PRs for #20 and #24, I'll let you know so you can test the full behavior of `setup.sh`. Thanks again!

**Comment by lucamlouzada** [comment]
Update: Now that #28  and #27 have been closed, I have merged these changes back to this branch. I also realized that `make_externals.sh` had inconsistencies between zsh and bash, so I updated it in 5aa7f6a.

@gentzkow , this should now be ready for your review. Let me know what you think so I can go ahead and open a PR. Thanks!

**Comment by gentzkow** [comment]
Awesome!

* Let's rename `/externals/` -> `/external/`
* We should add `/external/` to `.gitignore`
* We should comment out "/path/to/your/input/file.csv" in the default `get_inputs.sh` scripts so `setup.sh` doesn't produce a bunch of warnings the first time it is run

For a separate task: There is a lot of redundant code copied across the `get_inputs.sh` scripts. Can we abstract the common code to scripts in `/lib/` so the `get_inputs.sh` scripts are lighter weight?

**Comment by lucamlouzada** [comment]
> Let's rename /externals/ -> /external/

Done in a5a0ff2

> We should add /external/ to .gitignore

Done in 48e20ad

> We should comment out "/path/to/your/input/file.csv" in the default get_inputs.sh scripts so setup.sh doesn't produce a bunch of warnings the first time it is run

Done in 9080e07

> For a separate task: There is a lot of redundant code copied across the get_inputs.sh scripts. Can we abstract the common code to scripts in /lib/ so the get_inputs.sh scripts are lighter weight?

Great point. Will add to the to do list - we already have a to-do of doing the same with `make.sh`.

Thanks @gentzkow . Will create a PR for this issue

**Comment by lucamlouzada** [comment]
**Summary**
In this issue we created [`setup.sh`](https://github.com/gentzkow/GentzkowLabTemplate/blob/c50a1b9771699ef170dbeb87fe2e19963868c5e5/setup.sh) which lives in the root of the repo. It creates `local_env.sh`, runs `get_inputs.sh` in each module, and runs `make_externals.sh`. We also updated these other files to reflect the new setup script and ensure zsh and bash consistency. More details can be found in this issue thread and the associated PR #30.

Changes were merged to main in 26c6f30. Final state of the issue branch [here](https://github.com/gentzkow/GentzkowLabTemplate/tree/b100dae45d6b7bf2f40e8f9be715ccb6651f3914).
